testgrid/cmd/config/main.go

```diff
@@ -17,60 +17,201 @@ limitations under the License.
 package main
 
 import (
+	"context"
+	"errors"
 	"flag"
 	"fmt"
 	"io/ioutil"
+	"log"
+	"net/url"
 	"os"
 	"strings"
-)
+	"time"
+
+	"k8s.io/test-infra/testgrid/util/gcs"
 
-var (
-	yamlPaths  = flag.String("yaml", "", "comma-separated list of input YAML files")
-	printText  = flag.Bool("print-text", false, "print generated proto in text format to stdout")
-	outputPath = flag.String("output", "", "output path to save generated protobuf data")
+	"cloud.google.com/go/storage"
 )
 
-func errExit(format string, a ...interface{}) {
-	fmt.Fprintf(os.Stderr, format, a...)
-	os.Exit(1)
+type multiString []string
+
+func (m multiString) String() string {
+	return strings.Join(m, ",")
 }
 
-func main() {
+func (m *multiString) Set(v string) error {
+	*m = strings.Split(v, ",")
+	return nil
+}
+
+type options struct {
+	creds     string
+	inputs    multiString
+	oneshot   bool
+	output    string
+	printText bool
+}
+
+func gatherOptions() options {
+	o := options{}
+	flag.StringVar(&o.creds, "gcp-service-account", "", "/path/to/gcp/creds (use local creds if empty")
+	flag.BoolVar(&o.oneshot, "oneshot", false, "Write proto once and exit instead of monitoring --yaml files for changes")
+	flag.StringVar(&o.output, "output", "", "write proto to gs://bucket/obj or /local/path")
+	flag.BoolVar(&o.printText, "print-text", false, "print generated proto in text format to stdout")
+	flag.Var(&o.inputs, "yaml", "comma-separated list of input YAML files")
 	flag.Parse()
+	return o
+}
 
-	yamlFiles := strings.Split(*yamlPaths, ",")
-	if len(yamlFiles) == 0 || yamlFiles[0] == "" {
-		errExit("Must specify one or more YAML files with --yaml\n")
+func (o *options) validate() error {
+	if len(o.inputs) == 0 || o.inputs[0] == "" {
+		return errors.New("--yaml must include at least one file")
 	}
-	if !*printText && *outputPath == "" {
-		errExit("Must set --print-text or --output\n")
+
+	if !o.printText && o.output == "" {
+		return errors.New("--print-text or --output=gs://path required")
 	}
-	if *printText && *outputPath != "" {
-		errExit("Cannot set both --print-text and --output\n")
+	return nil
+}
+
+// announceChanges watches for changes to files and writes one of them to the channel
+func announceChanges(ctx context.Context, paths []string, channel chan []string) {
+	defer close(channel)
+	modified := map[string]time.Time{}
+	for _, p := range paths {
+		modified[p] = time.Time{} // Never seen
+	}
+
+	// TODO(fejta): consider waiting for a notification rather than polling
+	// but performance isn't that big a deal here.
+	for {
+		var changed []string
+		for p, last := range modified {
+			select {
+			case <-ctx.Done():
```

Comment:
This check seems unnecessary -- it's not a big deal if we read a few extra mtimes before exiting.

Reply:
We don't know how many --yaml files we are parsing? Seems easy enough to toss into the loop.